feat(contracts): add deploy configs and implementation deploy logic for kroma L1 contracts

sm-stack opened the pull request:
# Description

This PR does 2 things
- Add deploy configs for kroma L1 contracts
- Add `KromaDeployImplementation.s.sol`, which is to deploy implementation contracts of kroma L1.

## `KromaDeployImplementation`

### Background
The reason why I made this contract is, to make the future upstream process as easy as possible while not breaking the compatibility with future modifications on OP Stack. The current OP Stack highly depends on 2 script files when deploying the smart contract: `Deploy.s.sol` and `L2Genesis.s.sol`. So adding kroma-specific one in that level will break the actions at devnet & testing. 
For example, recently OP Stack made a breaking change on its [devnet structure](https://github.com/ethereum-optimism/optimism/commits/develop/kurtosis-devnet), which uses [kurtosis](https://docs.kurtosis.com/) instead of custom python file to initiate devnet. The `optimism-package` at kurtosis relies on `op-deployer`, and `op-deployer` calls `Deploy.s.sol` and `L2Genesis.s.sol` internally. 

So I thought that instead of adding a separate deploy script for Kroma contracts, integrating the deployment process of Kroma specific contracts inside `Deploy.s.sol` and `L2Genesis.s.sol` is much better. Please suggest any ideas if you have!

### Demonstration
The `KromaDeployImplementation.s.sol` contract takes the role of deploying the impl contracts, and doing the sanity checks of the deploy configs and deployment itself.

The `KromaDeployImplementation` contract will be initialized in `Deploy.s.sol`, and will be used to deploy kroma L1 contracts. There would be a function such as `deployKromaImplementations()`, which is similar with `deployImplementations` function. The full process will be like this:
1. `deployKromaImplementations()` function (which will be implemented in the following PR) will get the necessary constructor arguments from `DeployConfig.s.sol` and set it at `KromaDeployImplementationsInput` contract.
2. It will execute run() function at `KromaDeployImplementation`, which will then deploy the whole contracts and return it at the output format of `KromaDeployImplementationsOutput`.

## Considerations
- Optimism actively uses contract interfaces, so I had to add interfaces for governance contracts & some function interfaces for L1 contracts to read immutable variables. There is a cyclic dependency issue between ValidatorManager and AssetManager, so I couldn't add a check for ValidatorManager address at `assertValidAssetManager()` function.
- Optimism used `abi.encodeCall` and `<InterfaceName>.__constructor__` when deploying the impl contract, but I changed it to `abi.encodeWithSelector`. `abi.encodeCall` is a better one, since it checks the type of each variables when it is called, while `abi.encodeWithSelector` does not. The reason why I did like this was because ValidatorManager and AssetManager imports their own interfaces. If I add `__constructor__(...)` at the interface files, the contracts that imports those interfaces breaks, since there's no `__constructor__` function in their contract. Maybe removing interface imports from ValidatorManager and AssetManager will work, but not sure it's a good idea.

Please lmk if you have any idea to solve those problems.

sm-stack committed:
fix(contracts): revise the import version of OZ contracts at UpgradeGovernor

sm-stack committed:
feat(contracts): add interfaces of L1 governance-related contracts

sm-stack committed:
feat(contracts): add KromaDeployImplementation script and input format

sm-stack committed:
feat: add kroma specific deploy configs

sm-stack committed:
feat(contracts): add missing interfaces to kroma L1 contracts

sm-stack committed:
feat(contracts): add output setting and checking functions at KromaDeployImplementation.s.sol

sm-stack committed:
feat(contracts): add deploy logic of kroma specific L1 contracts

sm-stack committed:
chore(contracts): rename to KromaDeployImplementations.s.sol

sm-stack committed:
fix(contracts): fix incorrect constructor parameter at Colosseum

seolaoh commented on packages/contracts-bedrock/deploy-config/hardhat.json:
```diff
@@ -62,5 +62,37 @@
+  "validatorManagerTrustedValidator": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
+  "validatorManagerMinRegisterAmount": "0x32",
+  "validatorManagerMinActivateAmount": "0x64",
+  "validatorManagerMptFirstOutputIndex": "0x3",
```
We might not need this?
```suggestion
```

seolaoh commented on packages/contracts-bedrock/scripts/deploy/KromaDeployImplementations.s.sol:
```diff
@@ -0,0 +1,787 @@
+
+import { Constants } from "src/libraries/Constants.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { Bytes } from "src/libraries/Bytes.sol";
```
```suggestion
```

seolaoh commented on packages/contracts-bedrock/scripts/deploy/KromaDeployImplementations.s.sol:
```diff
@@ -0,0 +1,787 @@
+import { ITimeLock } from "interfaces/governance/ITimeLock.sol";
+import { IUpgradeGovernor } from "interfaces/governance/IUpgradeGovernor.sol";
+
+import { Blueprint } from "src/libraries/Blueprint.sol";
```
```suggestion
```

Pangssu commented on packages/contracts-bedrock/scripts/deploy/DeployConfig.s.sol:
```diff
@@ -176,6 +207,38 @@ contract DeployConfig is Script {
         useInterop = _readOr(_json, "$.useInterop", false);
+
+        // [Kroma: START]
+        validatorManagerTrustedValidator = stdJson.readAddress(_json, "$.validatorManagerTrustedValidator");
```
How about putting Kroma related configs inside the kroma object?
```
        validatorManagerTrustedValidator = stdJson.readAddress(_json, "$.kroma.validatorManagerTrustedValidator");
```

sm-stack replied:
Do you mean like declaring a struct like KromaConfigs and make a separate function to update the configs from the json at `read()` function here?

Pangssu replied:
well, ignore it please

seolaoh commented on packages/contracts-bedrock/scripts/deploy/KromaDeployImplementations.s.sol:
```diff
@@ -0,0 +1,787 @@
+    IZKProofVerifier internal _zkProofVerifierImpl;
+
+    function set(bytes4 _sel, address _addr) public {
+        require(_addr != address(0), "DeployImplementationsOutput: cannot set zero address");
```
```suggestion
        require(_addr != address(0), "KromaDeployImplementationsOutput: cannot set zero address");
```

seolaoh commented on packages/contracts-bedrock/scripts/deploy/KromaDeployImplementations.s.sol:
```diff
@@ -0,0 +1,787 @@
+        } else if (_sel == this.zkProofVerifierImpl.selector) {
+            _zkProofVerifierImpl = IZKProofVerifier(_addr);
+        } else {
+            revert("DeployImplementationsOutput: unknown selector");
```
```suggestion
            revert("KromaDeployImplementationsOutput: unknown selector");
```

seolaoh commented on packages/contracts-bedrock/scripts/deploy/KromaDeployImplementations.s.sol:
```diff
@@ -0,0 +1,787 @@
+    }
+
+    function assetManagerImpl() public view returns (IAssetManager) {
+        require(address(_assetManagerImpl) != address(0), "DeployImplementationsOutput: not set");
```
Why don't you write in this way same as `DeployImplementations.s.sol`?
```suggestion
        DeployUtils.assertValidContractAddress(address(_assetManagerImpl));
```

sm-stack committed:
fix(contracts): replace abi.encodeWithSelector with abi.encodeCall

sm-stack commented:
I've rebased this onto the target branch to apply the fix that removes importing interfaces at contracts (ValidatorManager, AssetManager). Also, based on that changes, I've managed to address the latter consideration, by [commit d0c6c59](https://github.com/kroma-network/optimism/pull/2/commits/d0c6c5986e32449a383ad133caac72f93794aafd).

+Addressed the first consideration by [the following commit](https://github.com/kroma-network/optimism/pull/2/commits/59407d2cc097374d6aaed672ac576ea2da5aa20a)!

sm-stack committed:
feat(contracts): add VALIDATOR_MANAGER function at IAssetManager

seolaoh commented on packages/contracts-bedrock/scripts/deploy/KromaDeployImplementations.s.sol:
```diff
@@ -0,0 +1,784 @@
+
+    function assertValidColosseum(KromaDeployImplementationsInput _dii) public view {
+        IColosseum colosseum = colosseumImpl();
+
```
Could you add this assertion for the contracts which call initalize() in constructor?
```suggestion
        DeployUtils.assertInitialized({ _contractAddress: address(colosseum), _slot: 0, _offset: 0 });

```

seolaoh commented on packages/contracts-bedrock/scripts/deploy/KromaDeployImplementations.s.sol:
```diff
@@ -0,0 +1,784 @@
+        _dio.checkOutput(_dii);
+    }
+
+    // -------- Deployment Steps --------
```
```suggestion
    // -------- Deployment Steps --------

    // --- Core Contracts ---
```

seolaoh commented on packages/contracts-bedrock/scripts/deploy/KromaDeployImplementations.s.sol:
```diff
@@ -0,0 +1,784 @@
+    }
+
+    function assetManagerImpl() public view returns (IAssetManager) {
+        require(address(_assetManagerImpl) != address(0), "DeployImplementationsOutput: not set");
```
Also apply it in below functions!
```suggestion
        require(address(_assetManagerImpl) != address(0), "KromaDeployImplementationsOutput: not set");
```

sm-stack committed:
feat: apply first PR review

sm-stack committed:
feat(contracts): update interface files with cast interface

sm-stack committed:
feat(contracts): remove interface imports from KromaMintableERC20, ProxyAdmin, and TokenMultiSigWallet

sm-stack committed:
feat(contracts): fix compile error and apply fmt at deploy scripts

sm-stack committed:
feat(contracts): exclude external and incomplete interfaces files from check

sm-stack committed:
chore(contracts): update snapshot and abi files

sm-stack commented:
Added 5 commits, doing the following things
- Thanks to @0xbenyun, interface files are generated by `cast interface` or `forge inspect`. Updated all added interfaces at [this commit](https://github.com/kroma-network/optimism/pull/2/commits/4d850ceb4c7f77bbe1af473992d06e6a3163e05d).
- Removed interface imports from `KromaMintableERC20` and `ProxyAdmin` contract, based on the interface policy of Optimism, at [this commit](https://github.com/kroma-network/optimism/pull/2/commits/d52a33164b879a70dfc4eff2a2d03677f9cc6338).
- The interface check script which is executed at `just build` generates error both at the incomplete interface files like `IL2OutputOracle`, and at newly imported OZ contracts. Thus, I added them to 'excluded contracts' at the interface checking script, at [this commit](https://github.com/kroma-network/optimism/pull/2/commits/8121cd725abb7000fd02264c607f87e84ff55fa6). Note that 3 contracts (`IL2OutputOracle` / `ProxyAdmin` / `IL1CrossDomainMessenger`) should be removed after resolving compile errors related to them at the tests.
- Added storage and abi files generated by `just snapshots-abi-storage` at [this commit](https://github.com/kroma-network/optimism/pull/2/commits/b47adf46acc1545d7885795da990f413440290dc).

seolaoh commented on packages/contracts-bedrock/src/universal/KromaMintableERC20.sol:
```diff
@@ -15,7 +15,7 @@ import { IKromaMintableERC20 } from "interfaces/universal/IKromaMintableERC20.so
 ///         Designed to be backwards compatible with the older StandardL2ERC20 token which was only
 ///         meant for use on L2.
-contract KromaMintableERC20 is IKromaMintableERC20, ERC20, ISemver {
+contract KromaMintableERC20 is ERC20, ISemver {
```
We can remove L10.

sm-stack replied:
It is used at the `supportInterface` function below, so removing it will generate a compile error.

dongchangYoo committed:
feat(packages): sycn L1CrossDomainMessenger to OP's (reordering import)

dongchangYoo committed:
feat(packages): use ProxyAdmin instead of Kroma's ProxyAdmin

dongchangYoo committed:
feat(packages): use OptimismPortal instead of KromaPortal

In the previous PR2, `KromaPortal` was replaced with `OptimismPortal`,
and this PR addresses and resolves the issues that arose during the rebase of that change.

kangsorang commented on packages/contracts-bedrock/scripts/checks/interfaces/main.go:
```diff
@@ -23,6 +23,14 @@ var excludeContracts = []string{
+	// External dependencies
+	"IERC5267", "IEIP712", "IVotesUpgradeable", "IAccessControlUpgradeable",
+	"IERC5192", "IERC721EnumerableUpgradeable",
+	// TODO: Temp. Remove the following line once the interface is fixed
+	"IL2OutputOracle", "IL1CrossDomainMessenger", "IProxyAdmin",
```
I think we've done enough work with Interface. Why don't we include this temporarily excluded part now?

kangsorang commented:
This pull request has been replaced by a new one that includes further changes.
